core/src/main/java/info/rmapproject/core/idservice/HttpArkIdService.java

```diff
@@ -0,0 +1,395 @@
+    private String userPassword = "";
+
+    /** Length of ID to validate against. */
+    private int idLength = -1;
+
+    /** String regex to validate an ID against. */
+    private String idRegex = "";
+
+    /**file for persistent reserved ID storage **/
+    private String idStoreFile = "/tmp/idStoreFile";
+    private static final String DATA = "ezidData";
+    /** the MVMap for storing the reserved EZIDs **/
+    private MVMap<Integer, String> ezids;
+    private int maxStoreSize=200;
+    /** indicates whether we have a process running which is replenishing the cache **/
+    private boolean replenishingCache = false;
+
+
+    /**
+     * Instantiates a new ARK ID service.
+     */
+    public HttpArkIdService() {
+
+    }
+
+    /* (non-Javadoc)
+     * @see info.rmapproject.core.idservice.IdService#isValidId(java.net.URI)
+     */
+    @Override
+    public boolean isValidId(URI id) throws Exception {
+        return isValidId(id.toASCIIString());
+    }
+
+
+    /**
+     * Check the string value of an ID is valid by checking it matches a regex and is the right length
+     *
+     * @param id the id
+     * @return boolean
+     * @throws Exception the exception
+     */
+    private boolean isValidId(String id) throws Exception {
+        boolean isValid = true;
+        if (idRegex!=null && idRegex.length()>0){
+            isValid = id.matches(idRegex);
+        }
+        if (isValid && idLength>0) {
+            isValid = (id.length()==idLength);
+        }
+        return isValid;
+    }
+
+    /* (non-Javadoc)
+     * @see info.rmapproject.core.idservice.IdService#createId()
+     */
+    public URI createId() throws Exception {
+        try {
+            return new URI(getEzid());
+        } catch (Exception e) {
+            throw new Exception("Failed to create a new ID.", e);
+        }
+    }
+
+    private synchronized String getEzid() throws Exception {
+        String id;
+        MVStore mvs = MVStore.open(idStoreFile);
+        ezids = mvs.openMap(DATA);
+        //if we are out of IDs, let's get some more from the EZID service endpoint
+        if(ezids.size() < 1){
+           if(!replenishingCache) {
+               getMoreEzids();
+           }
+        }
+
+        //let's make sure we have one to hand out. we can wait just a bit to let the previous
+        //call fill up the cache a bit
+        boolean shouldRetry;
+        int retryCounter = 0;
+        do{
+            retryCounter++;
+            shouldRetry = (retryCounter < maxRetryAttempts && ezids.size() == 0);
+			//WAIT FOR 5 SECS BEFORE RE-TRYING TO OVERCOME TEMPORARY NETWORK FAILURES
+			//OR THE EZID SERVER BEING BUSY SERVICING ANOTHER REQUEST.
+			if(shouldRetry) {
+                try {
+                    wait(RETRY_WAIT_TIME);
+                } catch (InterruptedException ie) {
+                    log.error("Wait interrupted in retry loop", ie);
+                }
+            }
+        } while (shouldRetry);
```

Comment:
This could loop forever.  Does it make sense to have an additional loop condition, say a timeout value?  If the timeout expires, exit the loop and throw a `RuntimeException`? (which I think you do below)

Reply:
@jrmartino does this loop still need a timeout?